Oxygen Generator Improvements

Jade-Harleyy opened the pull request:
This PR makes changes to oxygen generators that I believe will enhance the interactions available with them.

The first change adds the ability for players to find and change the oxygen generation rate either manually, or through wiring.

The second change adds a UI to oxygen generators that allows players to:
1. See if it is powered.
2. Manually change the oxygen generation rate.
3. See if the generation rate is being controlled through wiring.

The third change adds a new upgrade for oxygen generators that increases the oxygen generation rate.
This upgrade has a base cost of 2000 mk and 5 oxygenite shards, with tier restrictions of:
- Tier 1: Unavailable
- Tier 2: Level 1
- Tier 3: Level 2

Each level of this upgrade increases oxygen generation by 20%.

The intended effect of these changes is to add more depth to the mechanics of oxygen generators, with the end goal of getting players to engage with them beyond it just being a repairable device that fills oxygen tanks.

For example, traitors can set the generation amount to a lower value to silently starve the crew of oxygen.
Another example is players can create more complex "duty cycle" circuts to save fuel by dynamically changing the generation amount through wiring, therefore reducing the power consumption.

Jade-Harleyy committed:
Code changes and base content.

Jade-Harleyy committed:
Content changes.

Jade-Harleyy committed:
Change position of this.

JeroonBosch commented on Barotrauma/BarotraumaShared/LocalMods/oxygen-changes/Upgrades.xml:
```diff
@@ -0,0 +1,11 @@
+<UpgradeModules>
+  <UpgradeCategory identifier="oxygengenerators" items="oxygengenerator"/>
```
I believe these have no text description defined right now, unless I am missing something?

Jade-Harleyy replied:
I discovered unused localization tags in EnglishVanilla.xml, so it should be using those.

JeroonBosch replied:
Ah alright, yeah I didn't check for those
> Increase the amount of oxygen generated.
Maybe that should still be expanded to also mention the power consumption (which may be more relevant in most cases). But we could also take care of that.

Jade-Harleyy replied:
Maybe that should still be expanded to also mention the power consumption (which may be more relevant in most cases).

I've decided to remove the reduced power consumption for the reason provided in commit fdb553, so this is no longer neccessary.

JeroonBosch reviewed (commented):
Missing text for oxygengenerator upgrade?

itchyOwl reviewed (changes requested):
Overall looks pretty good, spotted only some minor code quality issues. Also would have to change the content definitions a bit, if we'd include the changes in the vanilla game.

Only skimmed through the xml changes, didn't go the UI code very thoroughly, as it's easier to spot the issues by testing, which I didn't do.

No objections on these changes overall.

JeroonBosch commented:
> For example, traitors can set the generation amount to a lower value to silently starve the crew of oxygen.

Without opening the UI of the oxygen generator, would this be visible in any way? At least, to me it would make sense if vents stopped turning when oxygen generation is at 0 (and slowing down at lower values, ideally), but also the generator itself no longer moving and/or showing a red light or orange light based on the percentage of oxygen generation.

Jade-Harleyy committed:
Address review comments.

Jade-Harleyy commented on Barotrauma/BarotraumaShared/LocalMods/oxygen-changes/filelist.xml:
```diff
@@ -0,0 +1,5 @@
+<ContentPackage name="oxygen-changes" modversion="1.0.0" gameversion="1.0.8.0">
```
Hey there, I'm not sure where I should place these files after addressing this, or how I should handle differences between changes/additions and mixed types.
Please let me know!

Jade-Harleyy committed:
Address review comments:
 - Serialize CurrFlow.
 - Only animate while producing oxygen.

Jade-Harleyy commented:
> Without opening the UI of the oxygen generator, would this be visible in any way? At least, to me it would make sense if vents stopped turning when oxygen generation is at 0 (and slowing down at lower values, ideally), but also the generator itself no longer moving and/or showing a red light or orange light based on the percentage of oxygen generation. 

Vents already stop turning when no oxygen is being produced, and both the vents and the generator make no sound when not producing anything.
I've also went ahead and stopped the generators from animating, though.

Jade-Harleyy committed:
Make this more accurate.

Jade-Harleyy committed:
Re-balance upgrades:
 - Remove decrease in power consumption; the improved generation lets players decrease the generation for less consumption anyway.
 - Decrease additional generation by 5%, for the same reason.

Jade-Harleyy committed:
Decrease the warning indicator sizes as well.

Jade-Harleyy committed:
Revert "Decrease the warning indicator sizes as well."
Turns out this applies to the original value (x - (40% + 40%)) instead of sequentially (x - 20% - 20%), leading to the new value being wrong.
This reverts commit 1708511eeeb9442ac854cb75187b0814ce3c76bf.

itchyOwl reviewed (approved):
Looks good to me from the technical perspective. Would have to decide what to do with this. As commented earlier, some changes are still required in the content definitions, if we'd include this in the vanilla game. But we can take care of that. For testing purposes, it's fine to have this implemented as a mod for now. GJ!

Regalis11 commented:
One concern I have about this is whether it's clear enough what's happening if someone for example turns the oxygen generation rate to 10%. How would a player know it's possible to adjust it, where is it communicated? How can a player tell the oxygen generation has been adjusted down, would it seem like a bug if they start suffocating for seemingly no reason?

mygamingaccount commented:
How is the sub supposed to handle grid overloads upon thrust deduction if the O2 genny is not there to suck up the extra power? Or is overvoltage not a thing any more in upcoming versions? (I'm cool with that by the way)

This change is very similar to [a circuit that I've designed](https://steamcommunity.com/sharedfiles/filedetails/?id=3104296897) for a campaign which involved having to manage fuel, but on that same sub I had to [entirely delete any overvoltage action](https://steamcommunity.com/sharedfiles/filedetails/?id=3104164209) just to make that kind of gameplay possible.

ubertpendragon commented:
> I think one thing that would alleviate the problem a bit is if the oxygen generation rate was visible on the generator and the vents somehow. Maybe the animation speed (and the volume of the sounds too?) could be relative to the oxygen generation rate?

I think that defeats the purpose of "silently" sabotaging the oxygen as Jade suggested

The use cases for O2 generator changes are:
-Starving fires (you previously had to disconnect the O2 power wire)
-Discreet Sabotage (instead of completely shutting off O2)
-Saving power

> The third change adds a new upgrade for oxygen generators that increases the oxygen generation rate.

This just reads as "reduce O2 power cost", subs don't really need more O2 generation since they're made with giving enough O2 to all rooms to begin with (it's not FTL where you can overpower a breach leaking oxygen with upgraded O2).
I believe O2 upgrades were removed because they were largely useless


Another problem I can think of with Saving Power is that you can survive on O2 shelves filling bottles. O2 shelves don't need the O2 generator to function since they're not LINKED but rather powered (sub builders wish they had been from the very start)

Regalis11 commented:
> I think that defeats the purpose of "silently" sabotaging the oxygen as Jade suggested

Possibly to some extent, but I think it's still important for there to be some kind of a cue that someone has sabotaged the oxygen. Maybe not something immediately obvious, but I feel like if we just add this sort of a "hidden feature" without communicating it anywhere in any way, it will just seem like a bug when the sub stops running out of oxygen for seemingly no reason or any clear way to fix it.

JoneKone commented:
> One concern I have about this is whether it's clear enough what's happening if someone for example turns the oxygen generation rate to 10%. How would a player know it's possible to adjust it, where is it communicated? How can a player tell the oxygen generation has been adjusted down, would it seem like a bug if they start suffocating for seemingly no reason?
> 
> I think one thing that would alleviate the problem a bit is if the oxygen generation rate was visible on the generator and the vents somehow. Maybe the animation speed (and the volume of the sounds too?) could be relative to the oxygen generation rate?

Or just vents are rotating slower than normal. Or both.
